flink-ml-lib/src/main/java/org/apache/flink/ml/classification/ftrl/Ftrl.java

```diff
@@ -0,0 +1,395 @@
+     */
+    public static class FtrlGlobalReducer implements ReduceFunction<DenseVector[]> {
+        @Override
+        public DenseVector[] reduce(DenseVector[] modelData, DenseVector[] newModelData) {
+            for (int i = 0; i < newModelData[0].size(); ++i) {
+                if ((modelData[1].values[i] + newModelData[1].values[i]) > 0.0) {
+                    newModelData[0].values[i] =
+                            (modelData[0].values[i] * modelData[1].values[i]
+                                            + newModelData[0].values[i] * newModelData[1].values[i])
+                                    / (modelData[1].values[i] + newModelData[1].values[i]);
+                }
+                newModelData[1].values[i] = modelData[1].values[i] + newModelData[1].values[i];
+            }
+            return newModelData;
+        }
+    }
+
+    /** Updates local ftrl model. */
+    public static class FtrlLocalUpdater extends AbstractStreamOperator<DenseVector[]>
+            implements TwoInputStreamOperator<
+                    Tuple2<Vector, Double>[], DenseVector[], DenseVector[]> {
+        private ListState<Tuple2<Vector, Double>[]> localBatchDataState;
+        private ListState<DenseVector[]> modelDataState;
+        private double[] n;
+        private double[] z;
+        private final double alpha;
+        private final double beta;
+        private final double l1;
+        private final double l2;
+        private DenseVector weights;
+
+        public FtrlLocalUpdater(double alpha, double beta, double l1, double l2) {
+            this.alpha = alpha;
+            this.beta = beta;
+            this.l1 = l1;
+            this.l2 = l2;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+
+            TypeInformation<Tuple2<Vector, Double>[]> type =
+                    ObjectArrayTypeInfo.getInfoFor(DenseVectorTypeInfo.INSTANCE);
+            localBatchDataState =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("localBatch", type));
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>("modelData", DenseVector[].class));
+        }
+
+        @Override
+        public void processElement1(StreamRecord<Tuple2<Vector, Double>[]> pointsRecord)
+                throws Exception {
+            localBatchDataState.add(pointsRecord.getValue());
+            alignAndComputeModelData();
+        }
+
+        @Override
+        public void processElement2(StreamRecord<DenseVector[]> modelDataRecord) throws Exception {
+            modelDataState.add(modelDataRecord.getValue());
+            alignAndComputeModelData();
+        }
+
+        private void alignAndComputeModelData() throws Exception {
```

Comment:
Would it be better to add some comments to this method, or divide this method further into several smaller methods? That could help improve readability.